Validate EnergyDispersion2D units

fabiopintore opened the pull request:
The PR solves a bug on the units of the input data for `~gammapy.irf.EnergyDispersion2D`, as reported in the issue #3725 .
It is now added a check on the units, and the code raises an error if the units are wrong.
The bug fix applies on `~gammapy.irf.core.py` although it is probably not the most elegant way to do it...

maxnoe commented on gammapy/irf/core.py:
```diff
@@ -45,6 +45,9 @@ def __init__(self, axes, data=0, unit="", meta=None, interp_kwargs=None):
             self.unit = data.unit
+            if self.axes.names == ["energy_true", "migra", "offset"] :
+                if self.unit != u.dimensionless_unscaled:
+                    raise ValueError(f"Error: {self.unit} is not an allowed unit. EDisp requires dimensionless data quantities!")
```
So I guess None of the IRFs specified units?

Then we should just add a `unit` attribute to all of the classes and use that here to check using:
```
if not self.unit.is_equivalent(data.unit):
    raise ValueError(...)
```

adonath replied:
Yes, this should be done in a general way. There is already a `IRF.unit` attribute, which specifies the currently used unit, bot not the expected one. I think the most elegant solution would be introduce a class level attribute `.default_unit` or similar, which specifies the expected unit equivalency for the data.

fabiopintore replied:
Thanks @maxnoe @adonath , I added the default units for all the IRF extensions (Edisp, Aeff, BKG and PSF) and I also modified the tests accordingly.
Unfortunately, I've read the last message only after pushing so I called the default units simply as `unit`...

adonath replied:
Yes, please adapt the name to `.default_unit` and add the check like `u.Unit(val).is_equivalent(self.default_unit)` here: https://github.com/gammapy/gammapy/blob/master/gammapy/irf/core.py#L145

maxnoe replied:
Shouldn't that setter transform any existing data?

At the moment (as is often the case when having this getter/setter setups) the user can easily create invalid data like this:

```
aeff = EffectiveArea2D(data=[[1, 2, 3, 4]] * u.m**2)
aeff.unit = u.cm**2  # I would expect that to *convert*, not *overwrite*
```

adonath replied:
Good point. Thinking about it again we should probably remove the setter completely? This is the behavior for `Quantity`:
```
from astropy import units as u
q = 5 * u.TeV
q.unit = "GeV"
```
Raises `AttributeError: can't set attribute`, which makes sense to me...

adonath replied:
Instead for maps we have: https://github.com/gammapy/gammapy/blob/master/gammapy/maps/core.py#L1506 Which mirrors a bit the `Quantity.to()` API.

fabiopintore replied:
Sorry,  it's not completely clear to me what's the final suggestion here, i.e. if I have to modify the `unit(self, val)` introducing some filters or not 😅.
For now, I've just implemented everywhere the simplest change (i.e. `default_unit`).

adonath replied:
The use cases to actually change the unit are more than rare, but it might still be reasonable to support it...

adonath replied:
@fabiopintore I think for this PR we probably keep as is, but we might need a follow up discussion to solve the question whether units should be settable at all (for `IRF` as well as `Map`). Right now they are which is fine, as it still a well define behavior. But I agree with @maxnoe that it is maybe unexpected and could cause more harm than benefit (especially because it is different behavior than for `Quantity`)

maxnoe replied:
If you *really* want to do it, you could do `map.quantity = u.Quantity(map.data, u.unit)` but then it's also clear that this is kind of dangerous.

So +1 for removing the unit setter (and probably even the data setter).

maxnoe commented on gammapy/irf/effective_area.py:
```diff
@@ -64,6 +64,7 @@ class EffectiveAreaTable2D(IRF):
 
     tag = "aeff_2d"
     required_axes = ["energy_true", "offset"]
+    unit = u.Unit("m**2")
```
Better use the defined units like this: `u.m**2`. No reason to go through string parsing.

maxnoe commented on gammapy/irf/edisp/core.py:
```diff
@@ -48,6 +48,7 @@ class EnergyDispersion2D(IRF):
 
     tag = "edisp_2d"
     required_axes = ["energy_true", "migra", "offset"]
+    unit = u.Unit("")
```
`u.one`

registerrier commented on gammapy/irf/psf/tests/test_table.py:
```diff
@@ -12,6 +14,21 @@ def psf_3d():
+    with pytest.raises(ValueError) as error:
+        psf3d = PSF3D(axes=[energy_axis, offset_axis, rad_axis],
+                 data=data)
+        assert error == (f"Error: {wrong_unit} is not an allowed unit. {psf3d.tag} requires dimensionless data quantities!")
```
is it dimensionless or per `sr`?

maxnoe replied:
No, a psf 3d should have unit `sr^-1`, https://gamma-astro-data-formats.readthedocs.io/en/latest/irfs/full_enclosure/psf/psf_table/index.html

registerrier reviewed (changes requested):
Thanks @fabiopintore . Checking units homogeneity is indeed important. 

There is some unclear. Why does the warning always states dimensionless quantity when the arguments list `Quantity` as a valid input. Do we want to only support `ndarray` + `Unit` initializer? If yes this should adapted accordingly.

maxnoe commented on gammapy/irf/core.py:
```diff
@@ -44,7 +44,10 @@ def __init__(self, axes, data=0, unit="", meta=None, interp_kwargs=None):
             self.data = data.value
-            self.unit = data.unit
+            if not self.default_unit.is_equivalent(data.unit):
+                raise ValueError(f"Error: {data.unit} is not an allowed unit. {self.tag} requires dimensionless data quantities!")
```
No, that is probably a copy/paste of the earlier version, it should be `{self.default_unit}` in the error message

fabiopintore replied:
Yes, indeed. It was a mistake. It was there left by the first version. I've fixed it!

fabiopintore commented:
Thank you @registerrier . The "dimensionless" word in the error message was a mistake due to a previous commit. I've corrected it and I've also modified some tests that were not good enough.
 
The only test that is not passing is on the `BKG3D` (`test_bkg_3d_wrong_units`), and I don't understand why. It mentions a `Did you mean to `re.escape()` the regex?` error, which I don't know how to avoid...

maxnoe commented:
@fabiopintore `match` uses a regular expression match to see if the error is what you expect. Many characters (e.g. `()`) have special meaning in regular expressions.

Since you define the exact error message, you should just compare the message, not use `match` or define a regex that is actually matching, e.g. 
```
assert error.match(f"Error: (.*) is not an allowed unit. (.*) requires (.*) data quantities.")```

fabiopintore commented:
Ah fantastic, I've got it. Very useful, @maxnoe

registerrier commented on gammapy/irf/tests/test_background.py:
```diff
@@ -220,6 +220,39 @@ def test_background_3d_read_gadf():
+    with pytest.raises(ValueError) as error:
+        Background3D(axes=[energy_axis, fov_lon_axis, fov_lat_axis],
+                 data=data)
+    assert error.match(f"Error: (.*) is not an allowed unit. (.*) requires (.*) data quantities.")
```
This probably does not need the f-string. Apparently this breaks the py flakes CI.

registerrier reviewed (approved):
Thanks @fabiopintore ! This looks good. I just left a small inline comment.

registerrier reviewed (approved):
Thanks @fabiopintore
